tests/test_models.py

```diff
@@ -1591,7 +1584,104 @@ def test_collapse_sets_false_with_user(self):
         self.node.collapse(user=self.user)
         assert_equal(self.node.is_expanded(user=self.user), False)
 
-
+    def test_next_descendants(self):
+        viewer = AuthUserFactory()
+        
+        root = ProjectFactory(creator=self.user)
+        comp1 = ProjectFactory(creator=self.user, parent=root)
+        comp1a = ProjectFactory(creator=self.user, parent=comp1)
+        comp1a.add_contributor(viewer, auth=self.consolidate_auth, permissions='read')
+        comp1b = ProjectFactory(creator=self.user, parent=comp1)
+        comp2 = ProjectFactory(creator=self.user, parent=root)
+        comp2.add_contributor(viewer, auth=self.consolidate_auth, permissions='read')
+        comp2a = ProjectFactory(creator=self.user, parent=comp2)
+        comp2a.add_contributor(viewer, auth=self.consolidate_auth, permissions='read')
+        comp2b = ProjectFactory(creator=self.user, parent=comp2)
+
+        descendants = root.next_descendants(
+            Auth(viewer),
+            condition=lambda auth, node: node.is_contributor(auth.user)
+        )
+        assert_equal(len(descendants), 2)  # two immediate children
+        assert_equal(len(descendants[0][1]), 1)  # only one visible child of comp1
+        assert_equal(len(descendants[1][1]), 0)  # don't auto-include comp2's children
+        
+    def test_get_descendants_recursive(self):
+        viewer = AuthUserFactory()
+        
+        root = ProjectFactory(creator=self.user)
+        comp1 = ProjectFactory(creator=self.user, parent=root)
+        comp1a = ProjectFactory(creator=self.user, parent=comp1)
+        comp1a.add_contributor(viewer, auth=self.consolidate_auth, permissions='read')
+        comp1b = ProjectFactory(creator=self.user, parent=comp1)
+        comp2 = ProjectFactory(creator=self.user, parent=root)
+        comp2.add_contributor(viewer, auth=self.consolidate_auth, permissions='read')
+        comp2a = ProjectFactory(creator=self.user, parent=comp2)
+        comp2a.add_contributor(viewer, auth=self.consolidate_auth, permissions='read')
```

Comment:
There's a lot of duplication across methods here. Would it make sense to add a separate test class with it's own setup?